src/workflow/get_mnemonic_passphrase.c

```diff
@@ -0,0 +1,156 @@
+#include "get_mnemonic_passphrase.h"
+
+#include <string.h>
+
+#include <hardfault.h>
+#include <ui/components/trinary_input_string.h>
+#include <ui/fonts/password_11X12.h>
+#include <ui/workflow_stack.h>
+
+#include "blocking.h"
+#include "confirm.h"
+#include "password_enter.h"
+#include "status.h"
+
+typedef enum {
+    MNEMONIC_PASSPHRASE_PW_ENTER,
+    MNEMONIC_PASSPHRASE_CONFIRM_MSG,
+    MNEMONIC_PASSPHRASE_CONFIRM_PASSPHRASE,
+    MNEMONIC_PASSPHRASE_TRY_AGAIN,
+    MNEMONIC_PASSPHRASE_FINISHED
+} mnemonic_passphrase_state_t;
+
+typedef struct {
+    mnemonic_passphrase_state_t state;
+    char* passphrase;
+    void (*callback)(char*, void*);
+    void* callback_param;
+} mnemonic_passphrase_data_t;
+
+/**
+ * Frees and zeros out the passphrase memory.
+ */
+static void _clean_passphrase(mnemonic_passphrase_data_t* data)
+{
+    if (data->passphrase) {
+        util_zero(data->passphrase, strlen(data->passphrase));
+        free(data->passphrase);
+        data->passphrase = NULL;
+    }
+}
+
+/**
+ * Invoked when the user has finished entering the passphrase.
+ */
+static void _passphrase_ready_cb(const char* passphrase, void* param)
+{
+    mnemonic_passphrase_data_t* data = (mnemonic_passphrase_data_t*)param;
+    data->passphrase = util_strdup(passphrase);
+    if (strlen(passphrase) == 0) {
+        // No need to confirm the empty passphrase.
+        data->state = MNEMONIC_PASSPHRASE_FINISHED;
+    } else {
+        data->state = MNEMONIC_PASSPHRASE_CONFIRM_MSG;
+    }
+}
+
+/**
+ * Invoked when the user has finished with the visual
+ * passphrase confirmation.
+ */
+static void _confirm_passphrase_cb(bool result, void* param)
+{
+    mnemonic_passphrase_data_t* data = (mnemonic_passphrase_data_t*)param;
+    if (!result) {
+        /* Confirmation failed. Start over. */
+        _clean_passphrase(data);
```

Comment:
nit: a little bit brittle, would not be needed if the passphrase was not on the heap but fixed size allocated with `SET_PASSWORD_MAX_PASSWORD_LENGTH`